### PR TITLE
[layer/node] Move input/output layers to LayerNode 

### PR DESCRIPTION
**Pull request by kparichay**
This patch moves input/output layers to LayerNode.
Also, create an interface for accessing/setting input/output layers
from in the LayerNode.

Refactor get/set number of inputs/outputs.
This is now handled by layer_node.
Updating the input/output layers updates the count.

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

**Review by lhs8928 — APPROVED** [review]
LGTM

File: nntrainer/graph/network_graph.cpp [hunk 1]
```diff
@@ -129,18 +124,8 @@ int NetworkGraph::realizeMultiInputType(
   std::shared_ptr<Layer> layer = lnode->getObject();
   graph.ensureName(*lnode, current.getName());
 
-  layer->setNumInputs(current.getNumInputs());
   lnode->setInputLayers(in_node->getInputLayers());
-  // layer->input_layers.clear();
-  // for (unsigned int i = 0; i < current.input_layers.size(); ++i)
-  //   layer->input_layers.push_back(current.input_layers[i]);
-
-  layer->setNumOutputs(current.getNumOutputs());
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:138]: no, outputs are managed at once when output layers are set.

File: nntrainer/graph/network_graph.cpp [hunk 2]
```diff
@@ -80,21 +80,22 @@ void NetworkGraph::updateConnectionName(const std::string &from,
     auto &layer = LNODE(node_list[i])->getObject();
     if (istrequal(layer->getName(), to))
       continue;
-    for (unsigned int j = 0; j < layer->input_layers.size(); ++j) {
-      if (istrequal(layer->input_layers[j], from)) {
-        layer->input_layers[j] = to;
-      }
-    }
+    LNODE(node_list[i])->updateInputLayers(from, to);
+    // for (unsigned int j = 0; j < layer->input_layers.size(); ++j) {
```
Comment (zhoonit) [inline nntrainer/graph/network_graph.cpp]: Unintended comment?

Comment (kparichay) [reply nntrainer/graph/network_graph.cpp]: this is already removed in a 2nd commit.

File: nntrainer/graph/network_graph.cpp [hunk 3]
```diff
@@ -353,36 +323,34 @@ void NetworkGraph::setOutputLayers() {
...
-    if (layer_idx->getNumOutputs() != layer_idx->output_layers.size()) {
-      if (layer_idx->output_layers.size() == 0) {
+    if (layer_idx->getObject()->getNumOutputs() != layer_idx->getNumOutputs()) {
+      if (layer_idx->getNumOutputs() == 0) {
         /** No output layer inplies its the last layer */
-        layer_idx->setNumOutputs(1);
-        layer_idx->output_layers.clear();
-        layer_idx->output_layers.push_back("__exit__");
+        layer_idx->setOutputLayers({"__exit__"});
+        ;
```
Comment (zhoonit) [inline nntrainer/graph/network_graph.cpp:352]: Additional semicolon?

File: nntrainer/graph/network_graph.cpp [hunk 4]
```diff
@@ -353,36 +323,34 @@ void NetworkGraph::setOutputLayers() {
...
           if (!already_exist)
-            layer_idx->output_layers.push_back(layer_i->getName());
+            layer_idx->addOutputLayers(layer_i->getName());
         }
       }
     }
 
-    if (layer_idx->getNumOutputs() != layer_idx->output_layers.size()) {
-      if (layer_idx->output_layers.size() == 0) {
+    if (layer_idx->getObject()->getNumOutputs() != layer_idx->getNumOutputs()) {
```
Comment (zhoonit) [inline nntrainer/graph/network_graph.cpp:348]: Could you check this? 

```suggestion
    if (layer_idx->getNumOutputs() != layer_idx->getNumOutputs()) {
```

File: nntrainer/layers/layer.cpp [hunk 5]
```diff
@@ -252,29 +252,6 @@ void Layer::setProperty(const PropertyType type, const std::string &value) {
       bias_initializer = (WeightInitializer)parseType(value, TOKEN_WEIGHT_INIT);
     }
     break;
-  case PropertyType::input_layers:
```
Comment (zhoonit) [inline nntrainer/layers/layer.cpp:255]: Nice work :)

File: nntrainer/layers/layer_node.h [hunk 6]
```diff
@@ -157,6 +158,105 @@ class LayerNode : public ml::train::Layer, public GraphNode {
...
+   * @brief this function helps exporting the layer in a predefined format,
+   * while workarounding issue caused by templated function type eraser
+   *
+   * @param exporter exporter that conatins exporting logic
+   * @param method enum value to identify how it should be exported to
+   */
+  virtual void
+  export_to(Exporter &exporter,
+            ExportMethods method = ExportMethods::METHOD_STRINGVECTOR) const {
+    exporter.saveResult(props, method, this);
```
Comment (zhoonit) [inline nntrainer/layers/layer_node.h:257]: Maybe call Layer::export_to in this stage?

File: nntrainer/utils/node_exporter.h [hunk 7]
```diff
@@ -207,11 +207,21 @@ class Layer;
 /**
  * @copydoc template <typename PropsType, typename NodeType> void
  * Exporter::saveTflResult(const PropsType &props, const NodeType *self);
+ * @todo remove this when layer is refactored
```
Comment (zhoonit) [inline nntrainer/utils/node_exporter.h:210]: This will be still needed, even though it might be an empty function.

Comment (kparichay) [reply nntrainer/utils/node_exporter.h:210]: as layer will become stateless, this won't be needed.

File: nntrainer/graph/network_graph.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:352]: removed.

File: nntrainer/utils/node_exporter.h (hunk 7 above)
Comment (zhoonit) [reply nntrainer/utils/node_exporter.h:210]: The reason is that default specialization is throwing operation. Supported functions (or operations) are whitelisted in this fashion.

File: nntrainer/graph/network_graph.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:348]: no, this is intended. `nntrainer::Layer` gives `getNumOutputs()` initialized with 1. input_dims/output_dims are always assumed to be at least 1. However, `nntrainer::LayerNode` relies on size of `input_layers/output_layers` to retrieve `getNumInputs/getNumOutputs` which can be zero. (when the layer is input/loss respectively).
This is a bit confusing for now but will be united once everything moves to layer context.

Comment (zhoonit) [reply nntrainer/graph/network_graph.cpp:348]: Ah, I see. Thanks for the explanation

**Review by zhoonit — APPROVED** [review]
LGTM

File: nntrainer/layers/layer_node.h (hunk 6 above)
Comment (kparichay) [reply nntrainer/layers/layer_node.h:257]: Updated with a new commit, as interpreter codes are also changed to use LayerNode than Layer.  Please check once.

Comment (zhoonit) [reply nntrainer/layers/layer_node.h:257]: Checked. Looks good to me :)

**Comment by jijoongmoon** [comment]
I think we need a description which explain which properties in LayerNode and Layer.. etc. in some where. ( Maybe wiki or as an Readme in layer dir or graph dir..)

**Review by jijoongmoon — APPROVED** [review]
LGTM. but why the ci is failed?

**Comment by kparichay** [comment]
> I think we need a description which explain which properties in LayerNode and Layer.. etc. in some where. ( Maybe wiki or as an Readme in layer dir or graph dir..)

This will be done post next week. Almost all the properties will move out to LayerNode. Once it is done, I will make the corresponding README/wiki.

File: nntrainer/utils/node_exporter.h (hunk 7 above)
Comment (kparichay) [reply nntrainer/utils/node_exporter.h]: done.

**Pull request merged**